# Coding Exercises

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Time Series Data in Pandas: Introduction


### Exercise 1: Intro to Time Series Data & DatetimeIndex

In [100]:
#run the cell!
import pandas as pd

1. __Import__ the csv-file "__aapl_fb.csv__" containing Stock Price Data for Apple and Facebook for the years 2015-2018! Convert the datatype in the column __Date__ to __datetime__ and set the column __Date__ as the __Index__ of the DataFrame! <br>
__Save__ the DataFrame in the variable __stocks__! __Fill in the gaps!__ 

In [101]:
stocks = pd.read_csv("aapl_fb.csv", sep = ',', index_col = "Date", parse_dates=['Date'])

How many Timestamps does __stocks__ contain? What was the __maximum stock price__ for Facebook in that Period?

In [102]:
# run the cell!
stocks

,AAPL,FB
Date,,
2015-01-02,109.330002,78.449997
2015-01-05,106.250000,77.190002
2015-01-06,106.260002,76.150002
2015-01-07,107.750000,76.150002
2015-01-08,111.889999,78.180000
...,...,...
2018-12-21,150.729996,124.949997
2018-12-24,146.830002,124.059998
2018-12-26,157.169998,134.179993


In [103]:
# run the cell!
stocks.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1005 entries, 2015-01-02 to 2018-12-28
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AAPL    1005 non-null   float64
 1   FB      1005 non-null   float64
dtypes: float64(2)
memory usage: 23.6 KB


In [104]:
# run the cell!
stocks.describe()

,AAPL,FB
count,1005.000000,1005.000000
mean,140.988338,133.415502
std,35.082112,35.846003
min,90.339996,74.050003
25%,112.120003,105.019997
50%,129.360001,129.070007
75%,168.630005,167.410004
max,232.070007,217.500000


There are in total ... 1005 Timestamps with stock price data. Facebooks´s maximum stock price was ... 217.50!

2. __Select__ the stock prices on __May 13, 2016__! What was the Price for __Apple__?

In [27]:
stocks.loc['2016-05-13']

AAPL     90.519997
FB      119.809998
Name: 2016-05-13 00:00:00, dtype: float64

The Apple Stock Price on that day was ... 90.52

3. __Select__ all stock prices in __September 2017__ and calculate the __mean__ for both stocks! What was the __mean stock price__ for __Apple__ in __September 2017__?

In [105]:
stocks.loc['2017-09'].mean()

AAPL    157.6085
FB      170.5910
dtype: float64

Apple´s monthly mean price for September 2017 is ... 157.61! 

4. __Create a Timestamp__ for August 10, 2019! (convert __2 alternative date strings__ of your choice) 

In [33]:
# Alt 1
pd.to_datetime('2019-08-10')


Timestamp('2019-08-10 00:00:00')

In [106]:
# Alt 2
pd.to_datetime('2019/08/10')

Timestamp('2019-08-10 00:00:00')

In [107]:

pd.to_datetime('10 August 2019')

Timestamp('2019-08-10 00:00:00')

5. __Change Frequency__ of stocks to __monthly__ ("Downsampling") and display the __last__ traded stock price per month! (Hint: The Index should show the last calender day of each month) <br>
What is the __last__ traded price for __Facebook__ in __March 2015__?

In [125]:
stocks.resample('ME').last()

,AAPL,FB
Date,,
2015-01-31,117.160004,75.910004
2015-02-28,128.460007,78.970001
2015-03-31,124.430000,82.220001
2015-04-30,125.150002,78.769997
2015-05-31,130.279999,79.190002
2015-06-30,125.430000,85.769997
2015-07-31,121.300003,94.010002
2015-08-31,112.760002,89.430000
2015-09-30,110.300003,89.900002


In [147]:
stocks_m=stocks.resample('ME').mean()

In [42]:
stocks.head()

,AAPL,FB
Date,,
2015-01-02,109.330002,78.449997
2015-01-05,106.250000,77.190002
2015-01-06,106.260002,76.150002
2015-01-07,107.750000,76.150002
2015-01-08,111.889999,78.180000


In [79]:
stocks.resample('ME').last()

,AAPL,FB
Date,,
2015-01-31,117.160004,75.910004
2015-02-28,128.460007,78.970001
2015-03-31,124.430000,82.220001
2015-04-30,125.150002,78.769997
2015-05-31,130.279999,79.190002
2015-06-30,125.430000,85.769997
2015-07-31,121.300003,94.010002
2015-08-31,112.760002,89.430000
2015-09-30,110.300003,89.900002


The last traded price is ... 82.22!

6. Calculate __monthly average/mean stock prices__ and save the new DataFrame in the variable __stocks_m__!

In [146]:
stocks_m=stocks.resample('ME').mean()

What was the average stock price for __Facebook__ in __April 2015__?

In [148]:
# run the cell!
stocks_m.head()

,AAPL,FB
Date,,
2015-01-31,110.641501,76.721500
2015-02-28,125.432105,76.951053
2015-03-31,125.970908,81.013183
2015-04-30,127.291428,82.091428
2015-05-31,128.761500,79.420000


The average price in april is ... 82.09!

7. __Downsample__ stocks to __weekly__ frequency, showing prices on __Fridays__!
What was __Facebook´s__ stock price on the __fourth Friday__ in __January 2015__?

In [149]:
stocks_m=stocks.resample('W-Fri').last()

Facebook´s stock price on the fourth Friday in January 2015 was ... 77.83!

In [150]:
stocks_m.head()

,AAPL,FB
Date,,
2015-01-02,109.330002,78.449997
2015-01-09,112.010002,77.739998
2015-01-16,105.989998,75.180000
2015-01-23,112.980003,77.830002
2015-01-30,117.160004,75.910004


# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Time Series Data in Pandas: Introduction


### Exercise 1: Intro to Time Series Data & DatetimeIndex

In [81]:
#run the cell!
import pandas as pd

1. __Import__ the csv-file "__aapl_fb.csv__" containing Stock Price Data for Apple and Facebook for the years 2015-2018! Convert the datatype in the column __Date__ to __datetime__ and set the column __Date__ as the __Index__ of the DataFrame! <br>
__Save__ the DataFrame in the variable __stocks__! __Fill in the gaps!__ 

In [82]:
stocks = pd.read_csv("aapl_fb.csv", parse_dates= ["Date"], index_col = "Date")

How many Timestamps does __stocks__ contain? What was the __maximum stock price__ for Facebook in that Period?

In [83]:
# run the cell!
stocks.head()

,AAPL,FB
Date,,
2015-01-02,109.330002,78.449997
2015-01-05,106.250000,77.190002
2015-01-06,106.260002,76.150002
2015-01-07,107.750000,76.150002
2015-01-08,111.889999,78.180000


In [84]:
# run the cell!
stocks.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1005 entries, 2015-01-02 to 2018-12-28
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AAPL    1005 non-null   float64
 1   FB      1005 non-null   float64
dtypes: float64(2)
memory usage: 23.6 KB


In [85]:
# run the cell!
stocks.describe()

,AAPL,FB
count,1005.000000,1005.000000
mean,140.988338,133.415502
std,35.082112,35.846003
min,90.339996,74.050003
25%,112.120003,105.019997
50%,129.360001,129.070007
75%,168.630005,167.410004
max,232.070007,217.500000


There are in total ... 1005 Timestamps with stock price data. Facebooks´s maximum stock price was ... 217.50!

2. __Select__ the stock prices on __May 13, 2016__! What was the Price for __Apple__?

In [86]:
stocks.loc["2016-05-13"]

AAPL     90.519997
FB      119.809998
Name: 2016-05-13 00:00:00, dtype: float64

The Apple Stock Price on that day was ... 90.52

3. __Select__ all stock prices in __September 2017__ and calculate the __mean__ for both stocks! What was the __mean stock price__ for __Apple__ in __September 2017__?

In [87]:
stocks.loc["2017-09"].mean()

AAPL    157.6085
FB      170.5910
dtype: float64

Apple´s monthly mean price for September 2017 is ... 157.61! 

4. __Create a Timestamp__ for August 10, 2019! (convert __2 alternative date strings__ of your choice) 

In [88]:
# Alt 1
pd.to_datetime("2019-08-10")

Timestamp('2019-08-10 00:00:00')

In [89]:
# Alt 2
pd.to_datetime("10 Aug 2019")

Timestamp('2019-08-10 00:00:00')

5. __Change Frequency__ of stocks to __monthly__ ("Downsampling") and display the __last__ traded stock price per month! (Hint: The Index should show the last calender day of each month) <br>
What is the __last__ traded price for __Facebook__ in __March 2015__?

In [90]:
stocks.resample("M").last() # old

C:\Users\bedla\AppData\Local\Temp\ipykernel_10992\1690053823.py:1: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  stocks.resample("M").last() # old


,AAPL,FB
Date,,
2015-01-31,117.160004,75.910004
2015-02-28,128.460007,78.970001
2015-03-31,124.430000,82.220001
2015-04-30,125.150002,78.769997
2015-05-31,130.279999,79.190002
2015-06-30,125.430000,85.769997
2015-07-31,121.300003,94.010002
2015-08-31,112.760002,89.430000
2015-09-30,110.300003,89.900002


In [91]:
stocks.resample("ME").last() # NEW (Pandas 3.x)

,AAPL,FB
Date,,
2015-01-31,117.160004,75.910004
2015-02-28,128.460007,78.970001
2015-03-31,124.430000,82.220001
2015-04-30,125.150002,78.769997
2015-05-31,130.279999,79.190002
2015-06-30,125.430000,85.769997
2015-07-31,121.300003,94.010002
2015-08-31,112.760002,89.430000
2015-09-30,110.300003,89.900002


The last traded price is ... 82.22!

6. Calculate __monthly average/mean stock prices__ and save the new DataFrame in the variable __stocks_m__!

In [93]:
stocks_m  = stocks.resample("M").mean() # NEW

C:\Users\bedla\AppData\Local\Temp\ipykernel_10992\3464350708.py:1: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  stocks_m  = stocks.resample("M").mean() # NEW


What was the average stock price for __Facebook__ in __April 2015__?

In [ ]:
# run the cell!
stocks_m.head()

The average price in april is ... 82.09!

7. __Downsample__ stocks to __weekly__ frequency, showing prices on __Fridays__!
What was __Facebook´s__ stock price on the __fourth Friday__ in __January 2015__?

In [99]:
stocks_last=stocks.resample("ME")
stocks_last.sum()

,AAPL,FB
Date,,
2015-01-31,2212.830017,1534.430002
2015-02-28,2383.209991,1462.069999
2015-03-31,2771.359984,1782.290016
2015-04-30,2673.119979,1723.919992
2015-05-31,2575.229995,1588.400008
2015-06-30,2811.749984,1832.240013
2015-07-31,2757.380006,2025.210005
2015-08-31,2381.290009,1927.350005
2015-09-30,2368.749995,1920.370011


Facebook´s stock price on the fourth Friday in January 2015 was ... 77.83!

# Well Done!

-------------------